ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -1103,6 +1253,12 @@ VALUE Message_encode(VALUE klass, VALUE msg_rb) {
   VALUE descriptor = rb_ivar_get(klass, descriptor_instancevar_interned);
   Descriptor* desc = ruby_to_Descriptor(descriptor);
 
+  if (desc->has_udts_transitively) {
```

Comment:
Yup, as noted in my top-level comment above, the serialize pre-pass isn't necessary, it was just the simplest way to get the initial prototype working. I'll fix this.